### DISCUSSION TITLE
LSP server for editor integration

### DISCUSSION THREAD

**Discussion by Nadrieril** [srid/neuron#647] [Ideas]
For best editor integration, it'd be awesome to have a LSP ([Language Server Protocol](https://code.visualstudio.com/api/language-extensions/language-server-extension-guide)) implementation for neuron. It could potentially provide autocompletion of links, automatically showing the link title next to a link, make links clickable, show backlinks, etc., and all this in an editor-agnostic way. That would greatly improve the experience of editing zettels with neuron.

**Comment by srid** [comment]
In theory one could also do things like new note *creation* (eg: https://github.com/felko/neuron-mode/issues/28) -- so all editors can just delegate to neuron in that aspect -- however I'm not sure if that's in the scope of LSP.

**Comment by sureyeaah** [comment]
I think we can have some kind of a hack here. Go-to-definition provides the mechanism for going to a different file and we could provide "note creation" as a codeaction.

**Comment by srid** [comment]
Some ideas: https://marketplace.visualstudio.com/items?itemName=sheilacat.zknotes

**Comment by NullSense** [comment]
Renaming of links could also be a functionality for an LSP. Go-to-definition as @sureyeaah mentioned of course. Now sure if views like backlinking would be implemented an an LSP level.

**Comment by Nadrieril** [comment]
Other suggestion: easy creation of a new note using current selection as the title

**Comment by srid** [comment]
Another: "refactor" by renaming a tag across the zettelkasten. 

cf. https://github.com/felko/neuron-mode/commit/beeb47fca65f78fbdaeaf216de85b6e8d7e6e71a

**Comment by aca** [comment]
@srid Are you working on this? or anyone?
I'm interested.

**Comment by srid** [comment]
@aca I'm not working on this. But @trevortknguyen was looking into this. Trevor, are you still interested? If so perhaps you two can collaborate? 

@aca You might want to [join Matrix](https://app.element.io/#/room/#neuron:matrix.org), the project channel.

**Comment by aca** [comment]
There will be an "official" neuron lsp implentation in haskell.
Until then, you might want to try https://github.com/aca/neuron-language-server.

**Comment by sureyeaah** [comment]
@aca I can help with the "official" neuron lsp!

**Comment by aca** [comment]
@sureyeaah. You should join [matrix](https://app.element.io/#/room/#neuron:matrix.org) and discuss about official neuron lsp.
For now, I will just play,experiment with the go lsp..

**Comment by trevortknguyen** [comment]
I do not think I will be able to work on the "official" neuron lsp.

**Comment by hoschi** [comment]
the call/caller graph features of LSP can also be utilized for viewing/navigating the z-index.

**Comment by srid** [comment]
An extension that displays neuron HTML preview to side would be pretty cool: https://twitter.com/owickstrom/status/1301241287727812610

**Comment by srid** [comment]
Reference implementation in Haskell: https://github.com/dhall-lang/dhall-haskell/tree/master/dhall-lsp-server

**Comment by srid** [comment]
If anybody feels like taking a stab at this, I added sample LSP server in Haskell here: https://github.com/srid/neuron/pull/539 

This provides a `neuron lsp` command that doesn't do much really, but you can use it with coc.vim in neovim using the below config. There is a hover handler which simply prints the notebook path.

```
$ cat ~/.config/nvim/coc-settings.json 
{
  "languageserver": {
    "neuron": {
      "command": "neuron",
      "args": ["lsp"],
      "filetypes": ["markdown"]
    }
  }
}
```

**Comment by srid** [comment]
One tiny benefit of using LSP would be that renames on large notebooks would be faster (`O(1)`). cf. https://github.com/svsool/vscode-memo/issues/286#issuecomment-796935882

**Comment by srid** [comment]
Inspiration: https://github.com/artempyanykh/zeta-note#screenshots

**Comment by samwalls** [comment]
Is there any other way to use neuron's searching facilities within vscode, currently? Ideally I'd like some quick command, mapped to a keyboard shortcut, where I can search zettel by their title, tag, or contents etc..

Regardless I might like to take a look at this, as I want to get familiar with serving LSP, writing utilities for zettel, and making "production-level" haskell. What's still left to do here? @srid

**Comment by srid** [comment]
You can try [task prompts](https://code.visualstudio.com/docs/editor/variables-reference#_input-variables) and then have vscode run the neuron command with prompted values.

For LSP, there is stub code here: https://github.com/srid/neuron/blob/master/src/Neuron/LSP.hs

I'd say the next steps are:

- Configure VSCode (this might require writing an extension in TypeScript?) to use `neuron lsp` (a command, hidden from `--help`) as the LSP server. In NeoVim, this is [straightforward to do](https://github.com/srid/neuron/issues/213#issuecomment-766440958).
- Confirm that the stub LSP.hs [hover message](https://github.com/srid/neuron/blob/210c2da158042e1158c417c4ff3a8430541b50c1/src/Neuron/LSP.hs#L44) works. This ensures that `neuron lsp` is hooked up to the editor successfully. 
- [Open neuron source in VSCode](https://github.com/srid/neuron/blob/master/CONTRIBUTING.md) and hack away the rest of the LSP features
    - Neuron's LSP.hs uses the [same library](https://hackage.haskell.org/package/lsp) as [haskell-language-server](https://github.com/haskell/haskell-language-server) (which provides IDE support for Haskell projects); so you can check the latter's source code for reference. The library's README also has some information: https://github.com/haskell/lsp/blob/master/README.md

As for "writing utilities for zettel" - I imagine that a lot of that can be done by `neuron query` in the command line, but if you can describe the specific workflows, I can give more info.

For contributing to production Haskell projects more generally, there are some [good first issues here](https://github.com/srid/neuron/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) in neuron's issue tracker.

Anyway, happy to help with any further specific questions. We also have a [Matrix chat room here](https://app.element.io/#/room/#neuron:matrix.org).

**Comment by mickael-menu** [comment]
[I've integrated an LSP server in `zk`](https://github.com/mickael-menu/zk), which is compatible with neuron. Just run `zk init` in your neuron notebook, then [follow the LSP instructions for your editor](https://github.com/mickael-menu/zk/blob/main/docs/editors-integration.md).

It supports neuron's tags and wiki-links.

(Kudos to @aca, your neuron LSP Go server got me started)

> * Configure VSCode (this might require writing an extension in TypeScript?) to use `neuron lsp`

That's right, you will need to create a VS Code extension for your LSP server. [I've created one for `zk` that is pretty minimal](https://github.com/mickael-menu/zk-vscode), which should be easy to modify to start `neuron lsp`.

One important bit is to configure the `activationEvents`. I chose to start my LSP client when a `.zk` folder is detected in the folder, you can probably do the same with `neuron.dhall`: https://github.com/mickael-menu/zk-vscode/blob/701cfd07db3529b039392067385fa81df594e74d/package.json#L14-L16
